xds/src/main/java/io/grpc/xds/MetadataRegistry.java

```diff
@@ -55,6 +62,49 @@ void removeParser(MetadataValueParser parser) {
     supportedParsers.remove(parser.getTypeUrl());
   }
 
+  /**
+   * Parses cluster metadata into a structured map.
+   *
+   * <p>Values in {@code typed_filter_metadata} take precedence over
+   * {@code filter_metadata} when keys overlap, following Envoy API behavior. See
+   * <a href="https://github.com/envoyproxy/envoy/blob/main/api/envoy/config/core/v3/base.proto#L217-L259">
+   *   Envoy metadata documentation </a> for details.
+   *
+   * @param metadata the {@link Metadata} containing the fields to parse.
+   * @return an immutable map of parsed metadata.
+   * @throws InvalidProtocolBufferException if parsing {@code typed_filter_metadata} fails.
+   */
+  public static ImmutableMap<String, Object> parseMetadata(Metadata metadata)
+      throws ResourceInvalidException, InvalidProtocolBufferException {
```

Comment:
Can this still throw `InvalidProtocolBufferException`? I could believe it, I just don't see where it happens.